### DISCUSSION TITLE
Go Formulae inconsistently build with -s -w

### DISCUSSION THREAD

**Discussion by FiloSottile** [Homebrew/discussions#723] [Everyday usage]
The default `std_go_args` don't include any `ldflags`, but many Formulae (over 100), go out of their way to add `"-ldflags", "-s -w"`, which omits symbol table and DWARF data. I am not aware of any case in which that could be required to get a working build. This breaks debugging, and anyway is inconsistent. Is there a homebrew-core policy on this, and should there be one?

**Comment by SMillerDev** [comment]
> I am not aware of any case in which that could be required to get a working build. This breaks debugging, and anyway is inconsistent.

We don't really package for debugging purposes. This setting probably helps with binary size though.

**Reply by FiloSottile** [reply]
It does help marginally with binary size. Is that something worth optimizing for?

Go builds with debug symbols by default (and doesn't make the flags to remove them easily accessible) because we think it's valuable for the something-went-wrong-to-debugging cycle to be short, without involving rebuilding or reinstalling.

In any case, I feel like they should be in `std_go_args`, or nowhere. Doing it per-Formula seems arbitrary.

**Reply by SMillerDev** [reply]
That's fair, could formula then define a second `ldflags` to set their version etc? If that's the case, could you make a pull request to add it to the standard go args?

**Reply by Loyalsoldier** [reply]
I like `-ldflags "-s -w -buildid="` more where `buildid` makes binaries exactly the same in the same platform and arch. In other words, one binary for all macOS.

**Reply by Bo98** [reply]
> could formula then define a second `ldflags` to set their version etc?

I think later `ldflags` completely override the prior ones: https://github.com/golang/go/issues/33581

**Reply by Bo98** [reply]
A reason we didn't include `-ldflags="-s -w"` to `std_go_flags` is that it doesn't actually work properly in all formulae - it actually breaks a few. The feeling was that `std` args should be applied everywhere, without exception. So if some formulae don't work with certain args, then those args shouldn't be a part of the "standard" set. Debian also omit `-s -w` from their standard args.

Unfortunately I don't have examples of formulae which break with `-ldflags="-s -w"` - I wasn't the maintainer who tried applying it to all formulae.

**Reply by FiloSottile** [reply]
> I like `-ldflags "-s -w -buildid="` more where `buildid` makes binaries exactly the same in the same platform and arch. In other words, one binary for all macOS.

Go builds should be identical even across cross compilation (when cgo is not involved, or the C compiler is well behaved), what differences in the buildid are you seeing?

In general, since `-s -w` is never required to make a build work, and especially if it sometimes breaks builds, I would avoid it everywhere, sticking to the standard Go build instructions most upstreams use and test. It costs some fairly compressible bytes on the wire, but it also has debugging value for users, who in Go don't generally expect binaries to be stripped.

**Reply by SMillerDev** [reply]
Binaries in brew aren't really made to be debugged by users though. I'd argue that the benefit for users with a slow internet connection outweighs the benefit for people who want to debug their install of a particular tool.

**Reply by Loyalsoldier** [reply]
> Go builds should be identical even across cross compilation (when cgo is not involved, or the C compiler is well behaved), what differences in the buildid are you seeing?

I try to build a Go project and you are right. Go v1.16 performs reproducible builds. I started to use `-buildid=(empty string)` in `-ldflags` because of https://github.com/golang/go/issues/34186

It seems it's unnecessary now in Go v1.16.
